### PR TITLE
Simple single client http multipart streaming server

### PR DESCRIPTION
**Pull request by fabito**
## Summary

Very simple example of a **single client** mjpeg streaming http server.

_There is no real reason for merging the PR. I used this as a learning exercise and we don't have a real use case for this example._

Uses `multipart/x-mixed-replace`:

```
HTTP/1.0 200 OK
Server: Imagr/1.0
Connection: Keep-Alive
Max-Age: 0
Expires: 0
Cache-Control: no-cache, private
Pragma: no-cache
Content-Type: multipart/x-mixed-replace;boundary=FRAME

--FRAME
Content-Type: image/raw
Content-Length: 123
Width: 324
Height: 324

<bytes>

--FRAME
Content-Type: image/raw
Content-Length: 123
Width: 324
Height: 324

<bytes>

--FRAME
Content-Type: image/raw
Content-Length: 123
Width: 324
Height: 324

<bytes>
```

## Tests

There is a webapp that consumes the stream and renders the frames in an html canvas

### CODE REVIEW

File: examples/camera_streaming_mjpeg/camera_streaming_mjpeg.cc [hunk 1]
```diff
@@ -0,0 +1,121 @@
...
+
+[[noreturn]] void Main() {
+  printf("MJPEG Streaming Example!\r\n");
+  // Turn on Status LED to show the board is on.
+  LedSet(Led::kStatus, true);
+
+  CameraTask::GetSingleton()->SetPower(true);
+  CameraTask::GetSingleton()->Enable(CameraMode::kStreaming);
+
+  const int server_socket = SocketServer(kPort, 10);
```
Comment (fabito) [inline examples/camera_streaming_mjpeg/camera_streaming_mjpeg.cc:94]: Using SocketServer so I can keep the TCP connection open to continuously write the captured jpegs into the http response.

File: examples/camera_streaming_mjpeg/camera_streaming_mjpeg.cc [hunk 2]
```diff
@@ -0,0 +1,121 @@
...
+
+
+void ProcessClient(int client_socket) {
+  // Check if the client socket is valid
+  if (client_socket < 0) {
+    printf("ERROR: Invalid client socket\r\n");
+    return;
+  }
+
+//   // Check if there is pending input
```
Comment (fabito) [inline examples/camera_streaming_mjpeg/camera_streaming_mjpeg.cc:40]: I am not getting any in information from the request at the moment, so not even bothering to read it...

File: examples/camera_streaming_mjpeg/camera_streaming_mjpeg.cc [hunk 3]
```diff
@@ -0,0 +1,121 @@
...
+  while (true) {
+    printf("INFO: Waiting for the client...\r\n");
+    const int client_socket = SocketAccept(server_socket);
+    if (client_socket == -1) {
+      printf("ERROR: Cannot connect client.\r\n");
+      continue;
+    }
+
+    printf("INFO: Client #%d connected.\r\n", client_socket);
+    ProcessClient(client_socket);
```
Comment (fabito) [inline examples/camera_streaming_mjpeg/camera_streaming_mjpeg.cc:109]: Only a single client is supported :-( 

I was thinking on using tasks to add multiple clients support. Maybe 3 tasks:

1. handles client connections to the webserver and add them to a `client_connections` list
2. handles getting picture frames from the camera and cache them locally
3. streams stored frames to all connected clients

Comment (Wal8800) [reply examples/camera_streaming_mjpeg/camera_streaming_mjpeg.cc:109]: What's the use case for having multiple client support?

Comment (fabito) [reply examples/camera_streaming_mjpeg/camera_streaming_mjpeg.cc:109]: Not really a practical one to be honest :-)
Just another learning exercise for inter tasks communication / synchronisation.

**Pull request closed without merging**